Lesson_Materials/In_Order_Queue/index.html

```diff
@@ -155,40 +151,74 @@
 					<div class="container">
 						<div class="col">
 							<code class="code-100pc"><pre>
-auto devicePtr   = usm_wrapper&lt;int&gt;(
-  malloc_device&lt;int&gt;(1024, inOrderQueue));
-  
-  inOrderQueue.memcpy(devicePtr, data, sizeof(int));
-  
-  inOrderQueue.parallel_for&lt;kernel_a&gt;(sycl::range{1024},
-  [=](sycl::id&lt;1&gt; idx){
-  devicePtr[idx] = /* some computation */
-});
-
-inOrderQueue.parallel_for&lt;kernel_b&gt;(sycl::range{1024},
-  [=](sycl::id&lt;1&gt; idx){
-  devicePtr[idx] = /* some computation */
-});
-
-inOrderQueue.memcpy(data, devicePtr, sizeof(int));
-
-<mark>inOrderQueue.wait();</mark>
+sycl:: queue Q{sycl::property::queue::in_order};
+for (auto task: tasks) {
+  Q.submit(task);
+}
+Q.wait();
+							</code></pre>
+						</div>
+						<div class="col" data-markdown>
+							* All events are executed serialy
+							* Ease of programing (no race condition)
+							* Zero concurency / Potential suboptimal performance
+						</div>
+					</div>
+				</section>
+				<section>
+					<div class="hbox" data-markdown>
+						#### Using an out-of-order queue (USM): But how to handle concurency
+					</div>
+					<div class="container">
+						<div class="col">
+							<code class="code-100pc"><pre>
+std::vector<sycl:: queue> Qs = map_queues_to_task();
+for (auto [Q,task]: queue_tasks) {
+  Q.submit(task);
+}
+for (auto [Q,task]: queue_tasks) {
+  Q.wait();
+}
 							</code></pre>
 						</div>
 						<div class="col" data-markdown>
-							* In the USM model dependencies are greatly simplified.
-							* It's no longer necessary to chain the commands together with `event` as they execute in the order they are enqueued.
-							* Simply calling `wait` on the the `queue` is sufficient.
+							* Manual Scheduling (painfull to extract full concurency)
+							* When correcly done, acheive concurency between queues;
```

Comment:
```suggestion
							* When correctly done, achieve concurrency between queues;
```
I sort of think that the focus on concurrency across queues is a little bit beyond the scope of this chapter. If we mention this it is also worth mentioning that multiple in order queues can be made from a single SYCL device. If we are running SYCL on a CUDA backend then we can "imagine" that these in order queues behave like cuda Streams. (However this is an implementation detail so best not to say that there is a definite mapping between SYCL in order queues and CUDA streams) :)

Reply:
This is precisely what the second exercise is doing (using multiple in-order queues).  Multiple in-order-queue and correct use of out-of-order queues (aka not calling `submit.wait()` all the time ... ) are critical optimizations for HPC. GPU are too big, it's rare that one kernel can saturate the full device.

And It's an "advanced" tutorial :)